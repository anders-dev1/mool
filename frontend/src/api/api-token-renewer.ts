import {AccessTokenClient, RenewAccessTokenCommand } from "./api-client";
import {store} from "../redux/store";
import {sessionActions} from "../redux/actions/session.action";

export async function RenewToken() {
    let refreshToken = localStorage.getItem("refresh-token");
    if (!refreshToken){
        return Promise.reject('There was no refresh token in storage.');
    }

    // use refresh token to get new access token
    let client = new AccessTokenClient();
    let request = new RenewAccessTokenCommand();
    request.refreshToken = refreshToken;

    await client.renew(request).then(
        result => {
            localStorage.setItem("access-token", result.accessToken);
        },
        error => {
            if (error.status === 404){
                store.dispatch(sessionActions.renewTokenExpired());
            }
        }
    );
}