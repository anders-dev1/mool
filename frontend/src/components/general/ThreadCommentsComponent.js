import React from "react";
import Divider from "./DividerComponent";
import DropDown from "./DropDownComponent";
import {commentOrderOptions} from "./__shared__/commentOrderOptions";
import {withTheme} from "styled-components";
import {connect} from "react-redux";
import {commentActions} from "../../redux/actions/comment.actions";
import {Comment} from "./Comment";
import Spinner from "./SpinnerComponent";
import {CommentCreation} from "./CommentCreationComponent";

class ThreadCommentsComponent extends React.PureComponent {
    constructor(props) {
        super(props);

        this.commentOrderChanged = this.commentOrderChanged.bind(this);
    }

    commentOrderChanged(order) {
        this.props.changeThreadCommentOrder(this.props.threadId, order.value);
    }

    render() {
        return (
            <div className={"ml1 mr1"}>
                <Divider className={"mt1 mb2"}/>
                <CommentCreation threadId={this.props.threadId}/>

                {this.props.comments?.length > 0 &&
                    <div>
                        <Divider className={"mt1 mb2"}/>
                        <div className={"flex justify-end"}>
                            <DropDown
                                right
                                options={commentOrderOptions}
                                initiallySelectedIndex={this.props.thread.commentOrder}
                                changedEvent={this.commentOrderChanged}/>
                        </div>
                    </div>
                }

                {this.props.thread.fetchingComments &&
                    <div data-testid={"fetchingCommentsSpinner"} className={"flex justify-center"}>
                        <Spinner size={"4"} color={"#007BFF"}/>
                    </div>
                }

                {
                    this.props.comments?.map((comment, index) => (
                        <div key={comment.id}>
                            <Comment threadId={this.props.threadId} commentId={comment.id} comment={comment}/>
                            {index !== this.props.comments.length - 1 && <Divider className={"mt1 mb2"}/>}
                        </div>
                    ))
                }
            </div>
        );
    }
}

function mapState(state, ownProps) {
    const thread = state.threadsReducer.threadsById[ownProps.threadId] ?? undefined;

    // Find comments for thread.
    const commentIds = state.commentsReducer.threadIdToCommentIds[ownProps.threadId];
    const comments = [];
    if (commentIds !== undefined){
        commentIds.forEach(commentId => {
            comments.push(state.commentsReducer.commentsById[commentId]);
        });
    }

    return {
        thread,
        comments
    }
}

const actionCreators = {
    updateCommentsForThread: commentActions.updateCommentsForThread,
    changeThreadCommentOrder: commentActions.changeThreadCommentOrder,
    addCommentToThread: commentActions.addCommentToThread
}

const connectedComponent = withTheme(connect(mapState, actionCreators)(ThreadCommentsComponent));
export {connectedComponent as ThreadCommentsComponent};
export * from './ThreadCommentsComponent';