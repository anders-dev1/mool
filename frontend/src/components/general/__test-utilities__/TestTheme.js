export const testTheme = {
    textColor: "black",
    primaryColor: "#007BFF",
    primaryHover: "rgba(0,123,255,0.33)",
    successColor: "#28A745",
    errorColor: "#F02849",
    backgroundColor: "#F0F2F5",
    buttonTextColor: "#FFFFFF",
    dividerColor: "#C7CBD1",
    inputColor: "#99A5B0",
    spinnerColor: "#FFFFFF transparent transparent transparent",
    font: "Acumin Pro, Arial, sans-serif !important;"
};