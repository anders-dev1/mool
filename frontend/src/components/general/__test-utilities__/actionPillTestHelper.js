import {screen} from "@testing-library/react";

export function findActionPillByIcon(icon) {
    return screen.getByText(icon).closest('div');
}

export function findActionPillNumberByIcon(icon) {
    const pill = findActionPillByIcon(icon);
    const numberP = pill.lastChild.lastChild;
    return numberP.textContent;
}