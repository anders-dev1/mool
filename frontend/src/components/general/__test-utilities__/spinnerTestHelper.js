export function findSpinner(container){
    return container.getElementsByClassName('lds-ring')[0];
}