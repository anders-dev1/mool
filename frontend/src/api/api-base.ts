import fetchIntercept, {FetchInterceptorResponse} from 'fetch-intercept';
import {RenewToken} from "./api-token-renewer";

class Request {
    url: string;
    config: Config;
    constructor(url: string, config: Config) {
        this.url = url;
        this.config = config;
    }
}

const ignoredPaths: string[] = [
    "/api/accesstoken/renew",
    "/api/accesstoken/authenticate"
]

export class ApiBase {
    constructor() {
        let originalRequest: Request;
        const interceptor = fetchIntercept.register({
            request(url: string, config: Config): Promise<any[]> | any[] {
                if (ignoredPaths.some((path) => url.endsWith(path))){
                    // Calling this function removes this interceptor from the list of interceptors.
                    // This stops our interceptor from processing any longer in the current request chain.
                    interceptor();
                    return [url, config];
                }

                originalRequest = new Request(url, config);

                const token = localStorage.getItem('access-token');
                if (!token){
                    return [url, config];
                }

                const myHeaders: HeadersInit = new Headers();
                myHeaders.set('Accept', 'application/json');
                myHeaders.set('Content-Type', 'application/json');
                myHeaders.set('Authorization', "Bearer " + token);

                config.headers = myHeaders;

                return [url, config];
            },
            response(response: FetchInterceptorResponse): Promise<FetchInterceptorResponse> | any {
                if (response.status != 401) {
                    return response;
                }

                if (originalRequest.url.includes('token')) {
                    return response;
                }

                let refreshToken = localStorage.getItem("refresh-token");
                if (!refreshToken) {
                    return response;
                }

                // use refresh token to get new access token
                return RenewToken().then(
                    result => {
                        const accessToken = localStorage.getItem('access-token');

                        const myHeaders: HeadersInit = new Headers();
                        myHeaders.set('Authorization', "Bearer " + accessToken);
                        myHeaders.set('Accept', 'application/json');
                        myHeaders.set('Content-Type', 'application/json');
                        originalRequest.config.headers = myHeaders;

                        return fetch(originalRequest.url, originalRequest.config).then(
                            result => {
                                return result;
                            }
                        );
                    }
                );
            }
        });
    }
}

export interface Config {
    method: string;
    headers: HeadersInit;
}