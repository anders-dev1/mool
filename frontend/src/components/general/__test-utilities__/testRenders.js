import React from 'react'
import {render} from '@testing-library/react'
import {Provider} from 'react-redux'
import {setupStore, store} from "../../../redux/store";
import {testTheme} from "./TestTheme";
import {ThemeProvider} from "styled-components";

// This is used to set up the store for tests.
export function renderWithProviders(
    ui,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        testStore = setupStore(preloadedState),
        ...renderOptions
    } = {}
) {
    function Wrapper({children}) {
        return(
            <Provider store={testStore}>
                <ThemeProvider theme={testTheme}>
                {children}
                </ThemeProvider>
            </Provider>
        );
    }

    return {testStore, ...render(ui, {wrapper: Wrapper, ...renderOptions})}
}

// This is used with the react-test-renderer to render dom in tests.
export function internalsForRenderer(
    ui,
    {
        preloadedState = {},
        testStore = setupStore(preloadedState)
    } = {}
){
    function Wrapper(children) {
        return(
            <Provider store={testStore}>
                <ThemeProvider theme={testTheme}>
                    {children}
                </ThemeProvider>
            </Provider>
        );
    }

    return Wrapper(ui);
}