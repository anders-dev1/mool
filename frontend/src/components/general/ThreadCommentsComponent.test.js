import {CommentOrder, CommentsClient} from "../../api";
import {internalsForRenderer, renderWithProviders} from "./__test-utilities__/testRenders";
import {ThreadCommentsComponent} from "./ThreadCommentsComponent";
import renderer from "react-test-renderer";
import {fireEvent, screen} from "@testing-library/react";

let commentsMock;
beforeEach(() => {
    commentsMock = jest.spyOn(CommentsClient.prototype, 'comments')
        .mockImplementation(() => Promise.resolve({comments: []}));
});
afterEach(() => {
    commentsMock.mockRestore();
});

const commentContent1 = "This is for testing1";
const commentContent2 = "This is for testing2";
const commentContent3 = "This is for testing3";
function createState(commentOrder = CommentOrder.OldestFirst, fetchingComments = false) {
    return {
        threadsReducer: {
            threadsById: {
                1: {
                    commentOrder: commentOrder,
                    postingComment: false,
                    numberOfComments: 0,
                    fetchingComments: fetchingComments
                }
            }
        },
        commentsReducer: {
            commentsById: {
                1: {
                    id: 1,
                    authorName: "Test Testerson",
                    content: commentContent1,
                    likes: 0,
                    likedByCurrentUser: false,
                    isChangingLikeStatus: false
                },
                2: {
                    id: 2,
                    authorName: "Test Testerson",
                    content: commentContent2,
                    likes: 0,
                    likedByCurrentUser: false,
                    isChangingLikeStatus: false
                },
                3: {
                    id: 3,
                    authorName: "Test Testerson",
                    content: commentContent3,
                    likes: 0,
                    likedByCurrentUser: false,
                    isChangingLikeStatus: false
                }
            },
            threadIdToCommentIds: {
                1: [1, 2, 3]
            }
        }
    }
}

function DropDownText(commentOrder) {
    if (commentOrder === CommentOrder.NewestFirst) {
        return "Newest first";
    } else {
        return "Oldest first";
    }
}

function renderComponent(commentOrder, fetchingComments) {
    renderWithProviders(
        <ThreadCommentsComponent threadId={1}/>,
        {preloadedState: createState(commentOrder, fetchingComments)});

    const dropdown = screen.getByText(DropDownText(commentOrder));

    return {dropdown};
}

// Needed for CommentCreationComponent.
function createNodeMock(element) {
    return document.createElement('div');
}

test('renders correctly', () => {
    const internalDom = internalsForRenderer(
        <ThreadCommentsComponent threadId={1}/>,
        {preloadedState: createState()});
    const tree = renderer.create(internalDom, {createNodeMock}).toJSON();
    expect(tree).toMatchSnapshot();
});

describe('comment order changed,', () => {
    test('when oldest first is selected, gets comments with the oldest first parameter', async () => {
        // Arrange
        const {dropdown} = renderComponent(CommentOrder.NewestFirst);

        // Act
        fireEvent.click(dropdown);
        await new Promise(process.nextTick);

        const oldestFirstOption = screen.getByText("Oldest first");
        fireEvent.click(oldestFirstOption);

        // Assert
        expect(commentsMock).toBeCalledWith(1, CommentOrder.OldestFirst);
    });

    test('when newest first is selected, gets comments with the newest first parameter', async () => {
        // Arrange
        const {dropdown} = renderComponent(CommentOrder.OldestFirst);

        // Act
        fireEvent.click(dropdown);
        await new Promise(process.nextTick);

        const newestFirstOption = screen.getByText("Newest first");
        fireEvent.click(newestFirstOption);

        // Assert
        expect(commentsMock).toBeCalledWith(1, CommentOrder.NewestFirst);
    });
});

describe('loading spinner', () => {
    test('when not fetching comments for thread, is not rendered', () => {
        // Arrange
        renderComponent(CommentOrder.OldestFirst, false);

        // Assert

        expect(screen.queryByTestId('fetchingCommentsSpinner')).toBeNull();
    });

    test('when fetching comments for thread, is rendered', async () => {
        // Arrange
        renderComponent(CommentOrder.OldestFirst, true);

        // Assert
        const component = screen.getByTestId('fetchingCommentsSpinner');
        expect(component).toHaveStyle('display: block');
    });
});

test('comments, when present in the store, are rendered', () => {
    // Arrange
    renderComponent(CommentOrder.OldestFirst, false);

    // Assert
    expect(screen.getByText(commentContent1)).toBeTruthy();
    expect(screen.getByText(commentContent2)).toBeTruthy();
    expect(screen.getByText(commentContent3)).toBeTruthy();
});
