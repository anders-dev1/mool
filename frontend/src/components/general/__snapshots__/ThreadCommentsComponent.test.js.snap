// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="ml1 mr1"
>
  <div
    className="sc-bczRLJ csXGiU mt1 mb2"
  />
  <form
    action="#"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="flex items-start"
    >
      <img
        className="sc-gKXOVf igLdOd"
        src="http://placekitten.com/256/256"
      />
      <div
        className="flex-auto ml1"
      >
        <div
          className="sc-evZas dXTKiT"
        >
          <textarea
            data-testid="commentCreationTextArea"
            name="comment"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Write your input here..."
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="flex justify-end mt1"
    >
      <button
        className="sc-ksZaOG blzxvH"
        data-testid="commentCreationPostBtn"
        type="submit"
      >
        <div
          className="sc-hAZoDl ciQoM"
          data-testid="childcontainer"
        >
          Post
        </div>
        <div
          className="sc-fnykZs gFyOxX"
          data-testid="spinner"
        >
          <div
            className="sc-iBkjds hEWUtX lds-ring"
            size="1.5"
          >
            <div
              className="sc-ftvSup kWpXHC"
              size="1.5"
            />
            <div
              className="sc-ftvSup kWpXHC"
              size="1.5"
            />
            <div
              className="sc-ftvSup kWpXHC"
              size="1.5"
            />
            <div
              className="sc-ftvSup kWpXHC"
              size="1.5"
            />
          </div>
        </div>
      </button>
    </div>
  </form>
  <div>
    <div
      className="sc-bczRLJ csXGiU mt1 mb2"
    />
    <div
      className="flex justify-end"
    >
      <div
        className="sc-gsnTZi jOVJit"
      >
        <div
          className="sc-dkzDqf kNuMDP"
          onClick={[Function]}
        >
          <div
            className="sc-hKMtZM VPycp"
          >
            <p>
              Oldest first
            </p>
            <img
              alt="expand comments"
              src="expand.svg"
              style={
                Object {
                  "transform": "translateY(-2px)",
                }
              }
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div>
      <div
        className="flex justify-start"
      >
        <img
          className="sc-gKXOVf igLdOd mr1"
          src="http://placekitten.com/256/256"
        />
        <div
          className="flex-column pt1"
        >
          <p
            className="sub"
          >
            Test Testerson
          </p>
          <p
            className="sub mt1"
            data-testid="commentContent"
          >
            This is for testing1
          </p>
        </div>
      </div>
      <div
        className="flex justify-end"
      >
        <div
          className="sc-papXJ iKXivi mr1"
          color="#007BFF"
          data-testid="commentLikeBtn"
          href="#"
          onClick={[Function]}
        >
          <p
            className="sc-jqUVSM gGcuMk"
          >
            üëç
          </p>
          <div
            className="sc-kDDrLX fDntCf"
          >
            <p
              className="sc-iqcoie dpfJlZ"
              data-testid="number"
            >
              0
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="sc-bczRLJ csXGiU mt1 mb2"
    />
  </div>
  <div>
    <div>
      <div
        className="flex justify-start"
      >
        <img
          className="sc-gKXOVf igLdOd mr1"
          src="http://placekitten.com/256/256"
        />
        <div
          className="flex-column pt1"
        >
          <p
            className="sub"
          >
            Test Testerson
          </p>
          <p
            className="sub mt1"
            data-testid="commentContent"
          >
            This is for testing2
          </p>
        </div>
      </div>
      <div
        className="flex justify-end"
      >
        <div
          className="sc-papXJ iKXivi mr1"
          color="#007BFF"
          data-testid="commentLikeBtn"
          href="#"
          onClick={[Function]}
        >
          <p
            className="sc-jqUVSM gGcuMk"
          >
            üëç
          </p>
          <div
            className="sc-kDDrLX fDntCf"
          >
            <p
              className="sc-iqcoie dpfJlZ"
              data-testid="number"
            >
              0
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="sc-bczRLJ csXGiU mt1 mb2"
    />
  </div>
  <div>
    <div>
      <div
        className="flex justify-start"
      >
        <img
          className="sc-gKXOVf igLdOd mr1"
          src="http://placekitten.com/256/256"
        />
        <div
          className="flex-column pt1"
        >
          <p
            className="sub"
          >
            Test Testerson
          </p>
          <p
            className="sub mt1"
            data-testid="commentContent"
          >
            This is for testing3
          </p>
        </div>
      </div>
      <div
        className="flex justify-end"
      >
        <div
          className="sc-papXJ iKXivi mr1"
          color="#007BFF"
          data-testid="commentLikeBtn"
          href="#"
          onClick={[Function]}
        >
          <p
            className="sc-jqUVSM gGcuMk"
          >
            üëç
          </p>
          <div
            className="sc-kDDrLX fDntCf"
          >
            <p
              className="sc-iqcoie dpfJlZ"
              data-testid="number"
            >
              0
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
